[F8N-755] Halt snapshots if supervisor's ledger is out-of-date

colinking opened the pull request:
If we were to lose snapshot data in S3, then new reflectors will have to start replaying the full DML ledger from the beginning to recreate an LDB. If this happened, then when the supervisor reboots and starts recreating its LDB, it'll start uploading quite out-of-date snapshots to S3 every 5m. After it uploads its first snapshot, any new reflectors that come online will take that snapshot and start rebuilding their LDB from that point-in-time. That would quickly lead to significant load on the ctldb.

In general, we never want to upload a snapshot unless it is relatively recent. This PR updates the supervisor to inspect its ledger latency when performing a snapshot. The exact limit that it enforces, the maximum ledger latency, is configurable, but will default to `10m`.

collinvandyck commented on pkg/supervisor/supervisor.go:
```diff
@@ -130,14 +142,33 @@ func (s *supervisor) Start(ctx context.Context) {
+			}
+			isAcceptableLatency := s.minimumLedgerLatency > latency
+			stats.Set("is_acceptable_latency", isAcceptableLatency)
+			stats.Set("latency", latency)
```
We already instrument ledger latency in the supervisor https://segment.datadoghq.com/dashboard/5hd-biq-s3m/ctlstore?from_ts=1571130225756&fullscreen_widget=362012904&live=true&tile_size=m&to_ts=1571144625756&fullscreen_section=overview so we can probably not worry about doing it here.

colinking replied:
Neat! Thanks

collinvandyck commented on pkg/supervisor/supervisor.go:
```diff
@@ -130,14 +142,33 @@ func (s *supervisor) Start(ctx context.Context) {
+			stats.Set("latency", latency)
+
+			if !isAcceptableLatency {
+				return nil
```
Would recommend logging here that we are skipping snapshotting the ldb, and include the latency (and threshold we compared it to) in the log message.

collinvandyck commented on pkg/cmd/ctlstore/main.go:
```diff
@@ -79,13 +79,14 @@ type executiveCliConfig struct {
+	ReflectorConfig      reflectorCliConfig `conf:"reflector" help:"reflector configuration"`
+	Shadow               bool               `conf:"shadow" help:"set this to true to emit shadow=true metric tags"`
+	Dogstatsd            dogstatsdConfig    `conf:"dogstatsd" help:"dogstatsd Configuration"`
+	MinimumLedgerLatency time.Duration      `conf:"minimum-ledger-latency" help:"Minimum ledger latency, at which the supervisor halts snapshots and waits for the ledger to catch up."`
```
Would `MaxPublishableLedgerLatency` be a better variable name here?

colinking replied:
🤦‍♂ should be `Maximum`! I tend to prefer more conciseness here, ala Go style :)

colinking commented:
Alright, confirmed in staging:

![image](https://user-images.githubusercontent.com/2907397/67052828-40b80900-f0f4-11e9-920c-a09305f2edc3.png)

Had to crank up the sample pct from 10% -> 100%, because with flagon performing caching, it barely hits ctlstore.

colinking commented on ldb_reader.go:
```diff
@@ -50,6 +50,23 @@ func (reader *LDBReader) GetLastSequence(ctx context.Context) (schema.DMLSequenc
+	GetLedgerLatency(ctx context.Context) (time.Duration, error)
+}
+
+type MockLatencier struct {
```
I'd rather not make this part of the public API. Would be curious where I might want to consider moving it to (a `pkg/latencier` package?).

colinking commented:
Added tests for the new supervisor behavior. To do so, I introduced an interface for `GetLedgerLatency` (which `LDBReader` already implements). This allows us to easily mock this in tests.

Open to feedback on that approach. Specifically, I don't think we should explicitly expose this mock, so we might want to move it to `pkg/ldb/latencier.go` or similar.

collinvandyck commented on pkg/cmd/ctlstore/main.go:
```diff
@@ -86,6 +86,7 @@ type supervisorCliConfig struct {
 	ReflectorConfig     reflectorCliConfig `conf:"reflector" help:"reflector configuration"`
 	Shadow              bool               `conf:"shadow" help:"set this to true to emit shadow=true metric tags"`
 	Dogstatsd           dogstatsdConfig    `conf:"dogstatsd" help:"dogstatsd Configuration"`
+	MaxLedgerLatency    time.Duration      `conf:"minimum-ledger-latency" help:"Minimum ledger latency, at which the supervisor halts snapshots and waits for the ledger to catch up."`
```
The `conf` tag here should be `maximum-ledger-latency` (or `min-ledger-latency` if you're into brevity). also the `help` tag should be adjusted similarly.

collinvandyck commented on pkg/supervisor/supervisor_test.go:
```diff
@@ -210,3 +225,72 @@ func TestSupervisorSnapshotReflectorCtl(t *testing.T) {
+
+	// Now clear the archive file.
+	err = os.Remove(archivePath)
+	if err != nil {
```
In the tests, let's try to use the stretchr test lib.  It makes error checking in tests much more concise, and it's what we use elsewhere in the codebase. Specifically:

```
require.NoErr(t, er)
```

collinvandyck commented on pkg/supervisor/supervisor_test.go:
```diff
@@ -15,10 +15,19 @@ import (
 )
 
+// Helper for building a GetLedgerLatency func.
+func mockGetLedgerLatency(duration time.Duration, err error) func(ctx context.Context) (time.Duration, error) {
```
We never use the `err` argument in any of the call sites.  An argument could be made to just remove the parameter here.

colinking replied:
Good point! I was thinking of adding tests for what happens when an `err` is returned, but never got around to that.
